ompi/mca/mtl/psm2/mtl_psm2_component.c

```diff
@@ -89,6 +93,30 @@ ompi_mtl_psm2_component_register(void)
 
     /* set priority high enough to beat ob1's default (also set higher than psm) */
     param_priority = 40;
+#if OPAL_CUDA_SUPPORT
+    /*
+     * If using CUDA enabled OpenMPI, the user likely intends to
+     * run with CUDA buffers. So, force-set the envvar here if user failed
+     * to set it.
+     */
+    cuda_env = getenv("PSM2_CUDA");
+    if (!cuda_env) {
+        opal_show_help("help-mtl-psm2.txt",
+                       "no psm2 cuda env", true,
+                       "not set",
+                       "Host buffers,\nthere will be a performance penalty"
+                       " due to OMPI force setting this variable now.\n"
+                       "Set environment variable to 0 if using Host buffers" );
+        setenv("PSM2_CUDA", "1", 0);
```

Comment:
Can't this lead to a segv if you put the "1" string into the environment, but then the PSM2 MTL is `dlclose()`d (because the PSM2 MTL was not selected)?

Reply:
It seems to be a harmless setting. I tested this by increasing OFI MTL priority to be higher than PSM2 and I did see the environment variable being set. But the application itself worked fine. Also- PSM2_DEVICES envvar setting is now done in component registration phase (PR #3834).

Reply:
I'm not 100% sure it's harmless.  I think you may be leaving now-unallocated memory in the `environ` array when the PSM MTL is unloaded.  It *probably* won't cause problems, but if anyone goes to try to read it -- especially if that virtual address is no longer valid -- it may cause problems.  You might be able to test that a little more rigorously (this is one reason that passing values around in the environment is kinda sucky).

If I'm right, it may be best to `strdup()` the `"1"`, even though that technically makes a minor memory leak.

And/or you might defer this setenv until you know that PSM2 is going to be used.  Then you will have a much smaller window for the bogus memory pointer in `environ` to cause a problem (i.e., because it won't become bogus until the PSM MTL is closed during `MPI_FINALIZE`, but that's usually on the way out of the process, anyway).